core/src/impl/Kokkos_ViewHooks.hpp

```diff
@@ -0,0 +1,466 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_VIEWHOOKS_HPP
+#define KOKKOS_VIEWHOOKS_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <functional>
+#include <memory>
+#include <type_traits>
+#include <map>
+#include <string>
+
+// default implementation for view holder specialization
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+template <class ViewType, class Enable = void>
+class ViewHookDeepCopy;
+
+template <class ViewType, class Enable = void>
+class ViewHookUpdate {
+ public:
+  static inline void update_view(ViewType &, const void *) {}
+  static constexpr const char *m_name = "ConstDefault";
+};
+
+template <class ViewType, class Enable = void>
+class ViewHookCopyView {
+ public:
+  static inline void copy_view(ViewType &, const void *) {}
+  static inline void copy_view(const void *, ViewType &) {}
+  static constexpr const char *m_name = "ConstDefault";
+};
+
+}  // namespace Impl
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Experimental {
+
+class ViewHolderBase {
+ public:
+  virtual size_t span() const                      = 0;
+  virtual bool span_is_contiguous() const          = 0;
+  virtual void *data() const                       = 0;
+  virtual void *rec_ptr() const                    = 0;
+  virtual std::string label() const noexcept       = 0;
+  virtual bool is_data_type_const() const noexcept = 0;
+
+  virtual ViewHolderBase *clone() const      = 0;
+  virtual size_t data_type_size() const      = 0;
+  virtual bool is_hostspace() const noexcept = 0;
+
+  // the following are implemented in specialization classes.
+  // View Holder is only a pass through implementation
+  // deep copy contiguous buffers
+  virtual void deep_copy_to_buffer(unsigned char *buff)   = 0;
+  virtual void deep_copy_from_buffer(unsigned char *buff) = 0;
+  // copy view contents that aren't contiguous
+  virtual void copy_view_to_buffer(unsigned char *buff)   = 0;
+  virtual void copy_view_from_buffer(unsigned char *buff) = 0;
+  // update view contents
+  virtual void update_view(const void *) = 0;
+  virtual ~ViewHolderBase()              = default;
+};
+
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Experimental {
+
+// ViewHolderCopy derives from ViewHolderBase and it
+// implement the pure virtual functions above.
+// the internal view has to be a copy so that downstream
+// containers can catch the result...
```

Comment:
```suggestion
// ViewHolderCopy derives from ViewHolderBase and it
// implement the pure virtual functions above.
// the view data member has to be a copy so that downstream
// containers can catch the result...
```